android/modules/ui/src/java/ti/modules/titanium/ui/WebViewProxy.java

```diff
@@ -110,9 +119,94 @@ public Object evalJS(String code)
 			Log.w(TAG, "WebView not available, returning null for evalJS result.");
 			return null;
 		}
+		if (callback != null) {
+			// When on Android 19+ we can use the builtin evalAsync method!
+			if (Build.VERSION.SDK_INT >= 19) {
+				final KrollFunction theCallback = callback;
+				view.getWebView().evaluateJavascript(code, new ValueCallback<String>() {
+					public void onReceiveValue(String value)
+					{
+						theCallback.callAsync(getKrollObject(), new Object[] { value });
+					}
+				});
+			} else {
+				// Just do our sync eval in a separate Thread.
+				Thread clientThread = new Thread(new EvalJSRunnable(view, getKrollObject(), code, callback),
+												 "TiWebViewProxy-" + System.currentTimeMillis());
+				clientThread.setPriority(Thread.MIN_PRIORITY);
+				clientThread.start();
+			}
+			return null;
+		}
+
+		if (KrollRuntime.getInstance().getKrollApplication().runOnMainThread()) {
+			Log.w(
+				TAG,
+				"Synchronous evalJS is not available when running on the main thread. Please supply an additional callback function argument to be invoked with the result as it's only parameter, to be called when the result is available.");
+			// FIXME This just times out on main thread! I don't know how else to get this to work...
+			// EvalJSSyncRunnable future = new EvalJSSyncRunnable(view, code);
```

Comment:
This code hung just as if I didn't spin off a thread, probably because I have to block below on the get call and the Javascript APIs have to use the UI thread to complete. In other words, I don't think we **can** run a sync version on main/ui thread.